### PR TITLE
Rewrite requester in OOP style and implement using of HTTP Agent

### PR DESCRIPTION
**Pull request by OlegPoberegets**

### CODE REVIEW

File: src/Requester.js [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+                throw new TypeError('config.agentOptions.maxSockets must be a positive integer or Infinity');
+            }
+
+            assertPositiveInteger(this._config.agentOptions.maxFreeSockets, 'config.agentOptions.maxFreeSockets');
+            assertPositiveInteger(this._config.agentOptions.timeout, 'config.agentOptions.timeout');
+        }
+
+        this._timeoutMsecs = this._config.timeoutMsecs || DEFAULT_REQUEST_TIMEOUT;
+        this._timing = this._config.timing || false;
+        this._agentOptions = this._config.agentOptions ? _.cloneDeep(this._config.agentOptions) : {};
```
Comment (WoZ) [inline src/Requester.js]: Why do you need to clone this one more time?

File: tests/Unit/Requester/constructor.js [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+    });
+
+    dataDriven(invalidMaxSocketsTypePayload, function () {
+        it('incorrect type of config.agentOptions.maxSockets, type = {type}', function (arg) {
+            assert.throws(
+                function () {
+                    new Requester({agentOptions: {maxSockets: arg.value}});
+                },
+                TypeError,
+                'config.agentOptions.maxSockets must be a positive integer'
```
Comment (WoZ) [inline tests/Unit/Requester/constructor.js]: constructor throws `config.agentOptions.maxFreeSockets must be a positive integer or Infinity`, so fix it pls and check why chai ignores mismatch.

File: tests/Unit/Requester/constructor.js [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+    ];
+
+    dataDriven(invalidMaxSocketsValuePayload, function () {
+        it('incorrect value of config.agentOptions.maxSockets, value = {type}', function (arg) {
+            assert.throws(
+                function () {
+                    new Requester({agentOptions: {maxSockets: arg.value}});
+                },
+                TypeError,
+                'config.agentOptions.maxSockets must be a positive integer'
```
Comment (WoZ) [inline tests/Unit/Requester/constructor.js]: constructor throws config.agentOptions.maxFreeSockets must be a positive integer or Infinity, so fix it pls and check why chai ignores mismatch.

**Comment by WoZ** [comment]
And remove please version number from package.json

File: src/Requester.js (hunk 1 above)
Comment (OlegPoberegets) [reply src/Requester.js]: Oh, It`s a mistake.

File: src/Requester.js [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+const {Agent} = require('http');
+const request = require('request');
+const {lookup} = require('dns-lookup-cache');
+
+const {assertResponse} = require('./ResponseAssert');
+const {RequestError} = require('./Error');
+
+const supportedIpFamily = 4;
+
+function promisifiedRequest(options) {
```
[the remark this answers is not in the document]
Comment (OlegPoberegets) [reply src/Requester.js]: With [request-promise-native](https://github.com/request/request-promise-native) we haven`t access to [http.ClientRequest](https://nodejs.org/dist/latest-v10.x/docs/api/http.html#http_class_http_clientrequest) events.

File: tests/Unit/Requester/constructor.js (hunk 2 above)
Comment (OlegPoberegets) [reply tests/Unit/Requester/constructor.js]: This test was deleted

File: tests/Unit/Requester/constructor.js (hunk 3 above)
Comment (OlegPoberegets) [reply tests/Unit/Requester/constructor.js]: This test was deleted

File: package.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "lcm-requester",
-  "version": "2.2.1",
+  "version": "3.0.0",
```
[the remark this answers is not in the document]
Comment (OlegPoberegets) [reply package.json]: OK

File: src/Requester.js [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+                    meta.remoteAddress = address;
+                }
+            });
+        });
+    });
+}
+
+function assertTimeout(timeoutMsecs, propertyName = 'timeoutMsecs') {
+    if (timeoutMsecs !== undefined && !Number.isSafeInteger(timeoutMsecs) || timeoutMsecs < 0) {
+        throw new TypeError(`${propertyName} must be a positive integer`);
```
[the remark this answers is not in the document]
Comment (OlegPoberegets) [reply src/Requester.js]: fixed

**Pull request merged**